File: openapi-client/scalaz/src/main/scala/org/openapitools/client/api/SendMailAdv.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import SendMailAdv._

case class SendMailAdv (
  /* The subject line of the email. */
  subject: String,
/* The email body.  If the string contains any HTML tags the message is automatically sent as `text/html`; otherwise it is sent as `text/plain`. */
  body: String,
from: EmailAddressTypes,
to: EmailAddressesTypes,
replyto: Option[EmailAddressesTypes],
cc: Option[EmailAddressesTypes],
bcc: Option[EmailAddressesTypes],
/* Optional list of file attachments.  Each file must be base64-encoded. Include `filename` so recipients see a meaningful attachment name. */
  attachments: Option[List[MailAttachment]],
/* Optional numeric ID of the mail order to send through.  If omitted the first active order on your account is used automatically.  Valid IDs are returned by `GET /mail`. */
  id: Option[Long])

object SendMailAdv {
  import DateTimeCodecs._

  implicit val SendMailAdvCodecJson: CodecJson[SendMailAdv] = CodecJson.derive[SendMailAdv]
  implicit val SendMailAdvDecoder: EntityDecoder[SendMailAdv] = jsonOf[SendMailAdv]
  implicit val SendMailAdvEncoder: EntityEncoder[SendMailAdv] = jsonEncoderOf[SendMailAdv]
}

File: openapi-client/zapier/models/SendMailAdv.js
const utils = require('../utils/utils');
const EmailAddressTypes = require('../models/EmailAddressTypes');
const EmailAddressesTypes = require('../models/EmailAddressesTypes');
const MailAttachment = require('../models/MailAttachment');

module.exports = {
    fields: (prefix = '', isInput = true, isArrayChild = false) => {
        const {keyPrefix, labelPrefix} = utils.buildKeyAndLabel(prefix, isInput, isArrayChild)
        return [
            {
                key: `${keyPrefix}subject`,
                label: `The subject line of the email. - [${labelPrefix}subject]`,
                required: true,
                type: 'string',
            },
            {
                key: `${keyPrefix}body`,
                label: `The email body.  If the string contains any HTML tags the message is automatically sent as `text/html`; otherwise it is sent as `text/plain`. - [${labelPrefix}body]`,
                required: true,
                type: 'string',
            },
            ...EmailAddressTypes.fields(`${keyPrefix}from`, isInput),
            ...EmailAddressesTypes.fields(`${keyPrefix}to`, isInput),
            ...EmailAddressesTypes.fields(`${keyPrefix}replyto`, isInput),
            ...EmailAddressesTypes.fields(`${keyPrefix}cc`, isInput),
            ...EmailAddressesTypes.fields(`${keyPrefix}bcc`, isInput),
            {
                key: `${keyPrefix}attachments`,
                label: `[${labelPrefix}attachments]`,
                children: MailAttachment.fields(`${keyPrefix}attachments${!isInput ? '[]' : ''}`, isInput, true), 
            },
            {
                key: `${keyPrefix}id`,
                label: `Optional numeric ID of the mail order to send through.  If omitted the first active order on your account is used automatically.  Valid IDs are returned by `GET /mail`. - [${labelPrefix}id]`,
                type: 'number',
            },
        ]
    },
    mapping: (bundle, prefix = '') => {
        const {keyPrefix} = utils.buildKeyAndLabel(prefix)
        return {
            'subject': bundle.inputData?.[`${keyPrefix}subject`],
            'body': bundle.inputData?.[`${keyPrefix}body`],
            'from': utils.removeIfEmpty(EmailAddressTypes.mapping(bundle, `${keyPrefix}from`)),
            'to': utils.removeIfEmpty(EmailAddressesTypes.mapping(bundle, `${keyPrefix}to`)),
            'replyto': utils.removeIfEmpty(EmailAddressesTypes.mapping(bundle, `${keyPrefix}replyto`)),
            'cc': utils.removeIfEmpty(EmailAddressesTypes.mapping(bundle, `${keyPrefix}cc`)),
            'bcc': utils.removeIfEmpty(EmailAddressesTypes.mapping(bundle, `${keyPrefix}bcc`)),
            'attachments': utils.childMapping(bundle.inputData?.[`${keyPrefix}attachments`], `${keyPrefix}attachments`, MailAttachment),
            'id': bundle.inputData?.[`${keyPrefix}id`],
        }
    },
}
